src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,590 @@
+
+This file defines positive and negative sets with respect to a vector measure and
+prove the Hahn decomposition (signed version).
+
+## Main definitions
+
+* `vector_measure.positive` : a set `i` is positive with respect to a vector measure if for all
+  measurable sets `j` contained in `i`, the measure of `j` is non-negative.
+* `vector_measure.negative` : a set `i` is negative with respect to a vector measure if for all
+  measurable sets `j` contained in `i`, the measure of `j` is non-positive.
+
+## Main results
+
+* `signed_measure.exists_disjoint_positive_negative_union_eq` : the Hahn decomposition theorem.
+* `signed_measure.exists_negative_set` : A measurable set of negative measure contains at least
+  one negative subset.
+
+## Tags
+
+positive, negative, Hahn decomposition
+-/
+
+noncomputable theory
+open_locale classical big_operators nnreal ennreal
+
+section lemmas
+
+variables {α : Type*}
+
+open set filter
+
+lemma exists_tendsto_Inf {S : set ℝ} (hS : ∃ x, x ∈ S) (hS' : ∃ x, ∀ y ∈ S, x ≤ y) :
+  ∃ (f : ℕ → ℝ) (hf : ∀ n, f n ∈ S), tendsto f at_top (nhds (Inf S)) :=
+begin
+  have : ∀ n : ℕ, ∃ t ∈ S, t < Inf S + 1 / (n + 1 : ℝ),
+  { exact λ n, (real.Inf_lt _ hS hS').1 ((lt_add_iff_pos_right _).2 nat.one_div_pos_of_nat) },
+  choose f hf using this,
+  refine ⟨f, λ n, (hf n).1, _⟩,
+  rw tendsto_iff_dist_tendsto_zero,
+  refine squeeze_zero_norm _ tendsto_one_div_add_at_top_nhds_0_nat,
+  intro n,
+  obtain ⟨hf₁, hf₂⟩ := hf n,
+  rw [real.dist_eq, real.norm_eq_abs, abs_abs,
+      abs_of_nonneg (sub_nonneg.2 (real.Inf_le S hS' hf₁))],
+  linarith,
+end
+
+lemma exists_tendsto_Sup {S : set ℝ} (hS : ∃ x, x ∈ S) (hS' : ∃ x, ∀ y ∈ S, y ≤ x) :
+  ∃ (f : ℕ → ℝ) (hf : ∀ n, f n ∈ S), tendsto f at_top (nhds (Sup S)) :=
+begin
+  have : ∀ n : ℕ, ∃ t ∈ S, Sup S - 1 / (n + 1 : ℝ) < t,
+  { intro n,
+    apply (real.lt_Sup _ hS hS').1,
+    rw [sub_lt, _root_.sub_self],
+    exact nat.one_div_pos_of_nat },
+  choose f hf using this,
+  refine ⟨f, λ n, (hf n).1, _⟩,
+  rw tendsto_iff_dist_tendsto_zero,
+  refine squeeze_zero_norm _ tendsto_one_div_add_at_top_nhds_0_nat,
+  intro n,
+  obtain ⟨hf₁, hf₂⟩ := hf n,
+  rw [real.dist_eq, real.norm_eq_abs, abs_abs, abs_of_nonpos],
+  { linarith },
+  { rw [sub_le, _root_.sub_zero],
+    exact real.le_Sup S hS' hf₁ },
+end
+
+end lemmas
+
+variables {α β : Type*} [measurable_space α]
+variables {M : Type*} [add_comm_monoid M] [topological_space M] [ordered_add_comm_monoid M]
+
+namespace measure_theory
+
+namespace vector_measure
+
+/-- A set `i` is positive with respect to a vector measure if for all
+measurable set `j`, `j ⊆ i`, `j` has non-negative measure. -/
+def positive (v : vector_measure α M) (i : set α) : Prop :=
+∀ ⦃j⦄, j ⊆ i → measurable_set j → 0 ≤ v j
+
+/-- A set `i` is negative with respect to a vector measure if for all
+measurable set `j`, `j ⊆ i`, `j` has non-positive measure. -/
+def negative (v : vector_measure α M) (i : set α) : Prop :=
+∀ ⦃j⦄, j ⊆ i → measurable_set j → v j ≤ 0
```

Comment:
Does this generalization make sense?
```suggestion
/-- With respect to a set `i`, a measure `v` is less than a measure `w` if
forall measurable set `j`, `j ⊆ i`, `v j ≤ w j`. -/
def le_on (i : set α) (v w : vector_measure α M) : Prop :=
∀ ⦃j⦄, j ⊆ i → measurable_set j → v j ≤ w j

/-- A set `i` is positive with respect to a vector measure if for all
measurable set `j`, `j ⊆ i`, `j` has non-negative measure. -/
def positive (v : vector_measure α M) (i : set α) : Prop :=
le_on 0 v

/-- A set `i` is negative with respect to a vector measure if for all
measurable set `j`, `j ⊆ i`, `j` has non-positive measure. -/
def negative (v : vector_measure α M) (i : set α) : Prop :=
le_on v 0
```

Reply:
With appropriate notation, you might be able to do away with `positive` and `negative` entirely, and use something like `0 ≤[i] v` and `v ≤[i] 0`  respectively.

Reply:
Or even, just define a `v.restrict i` which sends `vector_measure α M` to `vector_measure i M` (if such a definition can exist), and then you can write `v.restrict i ≤ 0`.

Reply:
I'll have to wait till the last push gets compiled to play around with this, but I think this is probably a good idea

Reply:
Which approach do you think is best? I'm leaning toward the second approach

Reply:
I have created a PR for defining the restriction of a vector measure: #8408

Reply:
Also #8410 introduces the missing `partial_order` instance

Reply:
@eric-wieser I have now refactored the file to use the new definition